### PR TITLE
Integration test experiment

### PR DESCRIPTION
**Pull request by kanishka-azimi**
@chreekat This has some essentials without the actual test suite(s)/runner. Let's discuss your preferences and advice for reorganizing.

I was ran using...
Terminal 1:
```
$ ./inttest-start-db.sh
```
Terminal 2:
```
$ ./inttest-start-webapp.sh
````

### CODE REVIEW

File: runtime-dirs/databrary-inttest/reload-gen-start-inttest [hunk 1]
```diff
@@ -0,0 +1,3 @@
+:module + Databrary.Main
+:main -w
+:main -c databrary.conf
```
Comment (kanishka-azimi) [inline runtime-dirs/databrary-inttest/reload-gen-start-inttest]: copy this from top level directory instead. (gen-start)

File: inttest-start-webapp.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+cp -R src/ app-databrary/ app-generate/ databrary.cabal runtime-dirs/databrary-inttest/
+mkdir -p runtime-dirs/databrary-inttest/config
+cp messages.conf volume.json runtime-dirs/databrary-inttest/    # copied from default.nix/databrary.nix
+cp -R web/ runtime-dirs/databrary-inttest/
+cp transcode transctl.sh runtime-dirs/databrary-inttest/
+mkdir -p runtime-dirs/databrary-inttest/solr/
+cp -r solr/solr.xml solr/log4j.properties solr/conf runtime-dirs/databrary-inttest/solr
+cat runtime-dirs/databrary-inttest/reload-gen-start-inttest | nix-shell --attr databrary-dev.env default.nix --command "cd runtime-dirs/databrary-inttest && ghci-databrary-inttest"
```
Comment (kanishka-azimi) [inline inttest-start-webapp.sh]: Last in separate script. Beginning called: copy-application

File: runtime-dirs/databrary-inttest/databrary.conf [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+   rotate = 100
+  }
+  access {
+   file = "stdout"
+   rotate = 100
+  }
+}
+db {
+  #host = "localhost"
+  sock = "../../databrary-nix-db/work/.s.PGSQL.5432"
```
Comment (kanishka-azimi) [inline runtime-dirs/databrary-inttest/databrary.conf]: Separate soon

File: inttest-start-db.sh [hunk 4]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+nix-shell --attr databrary-dev.env default.nix --command "./init-db"
```
Comment (kanishka-azimi) [inline inttest-start-db.sh]: Split into two scripts. Start/create server + user. Drop/create db.

**Comment by kanishka-azimi** [comment]
Why not use a thread to start web server or hspec-wai to issue a single request? The web application consists of background threads and external processes that are messy to startup separately at present. Should consider how we can make the other threads easier to launch intentionally from tests.

File: inttest-start-webapp.sh (hunk 2 above)
Comment (kanishka-azimi) [reply inttest-start-webapp.sh]: Delete and copy before each test suite run

File: runtime-dirs/databrary-inttest/databrary.conf [hunk 5]
```diff
@@ -0,0 +1,54 @@
+##See example.conf for complete conf options.
+secret = "bob"
```
Comment (kanishka-azimi) [inline runtime-dirs/databrary-inttest/databrary.conf]: Generate this and original config from m4 template

**Comment by kanishka-azimi** [comment]
Single test:
```
./inttest-start-db.sh &
./inttest-start-webapp.sh &
res=$(runghc testcase1.hs)
kill %1
kill %2
exit $res
```

testcase1.hs
```
main = testMain (testGroup "" [testCase])

testCase :: TestCase
testCase =
  test "t1" (do
     -- json <- parseResult <$> (http call)
     json @?= expectedJson)
```

File: runtime-dirs/databrary-inttest/databrary.conf (hunk 5 above)
Comment (kanishka-azimi) [reply runtime-dirs/databrary-inttest/databrary.conf]: linked instead

File: runtime-dirs/databrary-inttest/databrary.conf (hunk 3 above)
Comment (kanishka-azimi) [reply runtime-dirs/databrary-inttest/databrary.conf]: since using a relative path, linking to the original config worked

**Comment by chreekat** [comment]
I'm going to wait on thinking about integration tests until we've made 2 weeks of progress on unit tests.

fwiw, our functional tests *are* integration tests, at the level that matters most (i.e. the highest level), and they seem to be pretty complete. So I am not especially worried about integration tests right now.

**Pull request closed without merging**